### PR TITLE
Updates Expansion Service Container to support upgrading using the schema-transform ID

### PR DESCRIPTION
**Pull request by chamikaramj**
Also includes several other fixes needed to expand Managed Iceberg I/O using the Java Expansion Service container.

This fixes https://github.com/apache/beam/issues/31430.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by chamikaramj** [comment]
R: @ahmedabu98

File: sdks/java/io/iceberg/build.gradle [hunk 1]
```diff
@@ -57,8 +57,8 @@ dependencies {
 
     testImplementation library.java.hadoop_client
     testImplementation library.java.bigdataoss_gcsio
-    testImplementation "com.google.cloud.bigdataoss:gcs-connector:hadoop2-2.2.16"
-    testImplementation "com.google.cloud.bigdataoss:util-hadoop:hadoop2-2.2.16"
+    testImplementation library.java.bigdataoss_gcs_connector
+    testImplementation library.java.bigdataoss_util_hadoop
```
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/java/io/iceberg/build.gradle:61]: I'm just adding these dependencies to the top level in this PR so dependencies should be exactly the same. But I can try triggering as well.

[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/java/io/iceberg/build.gradle:61]: I mean, I'm adding "hadoop2-2.2.16" to the top level. I don't think there's a "2.2.16" version for these jars.

[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/iceberg/build.gradle:61]: There's a slight difference in the dependencies though. This essentially changes the dependencies from `com.google.cloud.bigdataoss:...:hadoop2-2.2.16` to `com.google.cloud.bigdataoss:...:2.2.16`

Noting that the latter didn't work with me when I was writing integration tests

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/construction/TransformUpgrader.java [hunk 2]
```diff
@@ -184,26 +203,33 @@ RunnerApi.Pipeline updateTransformViaTransformService(
...
+    }
 
     RunnerApi.PTransform.Builder ptransformBuilder =
         RunnerApi.PTransform.newBuilder()
             .setUniqueName(transformToUpgrade.getUniqueName() + "_external")
             .setSpec(
                 RunnerApi.FunctionSpec.newBuilder()
                     .setUrn(transformToUpgrade.getSpec().getUrn())
-                    .setPayload(ByteString.copyFrom(payload.toByteArray()))
+                    .setPayload(ByteString.copyFrom(payloadBytes))
```
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/construction/TransformUpgrader.java:232]: Yes. This is the protocol for the ExpansionRequest, the payload depends on the URN.

If the URN is "BeamUrns.getUrn(SCHEMA_TRANSFORM)" payload should be "SchemaTransformPayload".

https://github.com/apache/beam/blob/c90fd8320a571573e9375318818bdb1a7c15d421/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/external_transforms.proto#L57

**Comment by chamikaramj** [comment]
Iceberg tests are passing:

https://github.com/apache/beam/actions/runs/9308308071/job/25621440114?pr=31451
https://github.com/apache/beam/actions/runs/9308308087/job/25621440320?pr=31451

**Comment by chamikaramj** [comment]
Also added a unit test that cover's TransformUpgrader changes here.

ExpansionService changes will need an integration test (a separate PR).

PTAL.

**Review by ahmedabu98 — APPROVED** [review]
LGTM

**Comment by chamikaramj** [comment]
Failures seems to be unrelated.

File: sdks/java/io/expansion-service/build.gradle [hunk 3]
```diff
@@ -43,6 +43,11 @@ dependencies {
   permitUnusedDeclared project(":sdks:java:io:kafka") // BEAM-11761
   implementation project(":sdks:java:io:kafka:upgrade")
   permitUnusedDeclared project(":sdks:java:io:kafka:upgrade") // BEAM-11761
+
+  // Needed by Iceberg I/O users that use GCS for the warehouse location.
+  implementation library.java.bigdataoss_gcs_connector
+  permitUnusedDeclared library.java.bigdataoss_gcs_connector
```
Comment (ahmedabu98) [inline sdks/java/io/expansion-service/build.gradle:49]: Should we add this dependency to `java/io/iceberg` instead?

**Pull request merged**